Schema

tombentley opened the pull request:
### Type of change

- Enhancement / new feature

### Description

This PR adds support for generating different schemas for different versions to the `CrdGenerator` and makes the CLI of that tool more flexible.

* `@Crd.spec.version` is removed. It's redundant since it's always the first item in the list of `@Crd.versions` (apiserver enforces this).
* `@Crd.apiVersion` (which is the version of the CRD API which the generated CRD will have, and nothing to do with the versions of the Kafka CRD which will be generated) is replaced with a CLI option.
* The CLI gains options for specifying which CR API versions to include in the generated CRD. This is useful primarily for testing and manual experimentation, e.g. it allows to generate a CRD with just `v1beta1` and `v1beta2` but not `v1alpha1`.
* The CLI gains an option for specifying the target Kubernetes version range. The intention here is to catch things like `v1` CRD targetting Kube 1.11, or `v1beta1` targetting some future version of Kube whith that CRD API version removed.
* Classes to represent Kubernetes and API versions and version ranges are added.
* The existing `@Pattern`, `@Maximum`, `@Minimum` and `@Description` annotations are given `apiVersions` methods and made `@Repeatable`. This allows for different constraints in different versions.
* The existing type `@Alternative` is given an `apiVersions` method to allow different type alternatives in different API verions. 
* A new `@PresentInVersions` annotation is added for being able to add and remove properties in specific versions.

Together these changes enable us to start working on a `v1beta2` version of `Kafka`, but still publishing only `v1alpha1`, `v1beta`.

It's worth pointing out a couple of tests which I wrote in the course of this:

1. `StructuralCrdIT` checks that the gestating `v1beta2` is structural.
2. `ApiEvolutionCrdIT` tests how we might evolve the schema over a number of Strimzi releases. 0.20 already adds `listeners` (which is not structural), a future `0.2x` might add `v1beta2`, but keep `v1beta1` as the stored version, and then a `0.2y` (y>x) might make `v1beta2` the stored version. Finally `0.2z` (z>y) might remove `v1beta1` and use CRD `v1` API.

samuel-hawker commented on api/src/test/java/io/strimzi/api/kafka/model/ApiEvolutionCrdIT.java:
```diff
@@ -0,0 +1,232 @@
+    private void assertV1beta2CreateFailure(String name) {
+        try {
+            LOGGER.info("Check can't create map-listener via v1beta2");
+            v1beta2Create(name, mapListener(), null);
```
```suggestion
            assartThrows(() -> v1beta2Create(name, mapListener(), null))
```

see-quick commented on api/src/test/java/io/strimzi/api/kafka/model/ApiEvolutionCrdIT.java:
```diff
@@ -0,0 +1,232 @@
+        return Crds.kafkaV1Beta2Operation(cluster.client().getClient()).inNamespace(NAMESPACE);
+    }
+
+    private GenericKafkaListener listListener() {
```
Wouldn't be better to name it createInternalPlainListener or am I missing something?

tombentley replied:
The precise details of the listener don't matter (the CR is never actually consumed by the operator), instead what we're testing here is that we API server lets us create a Kafka CR with listeners as a list rather than a map.

see-quick commented on crd-generator/src/main/java/io/strimzi/crdgenerator/CrdGenerator.java:
```diff
@@ -652,44 +898,91 @@ ArrayNode stringArray(Iterable<String> list) {
+                if (arg.startsWith("--")) {
+                    if (arg.equals("--yaml")) {
+                        yaml = true;
+                    } else if (arg.equals("--label")) {
```
Just curious, but wouldn't be better to use `switch` construction for the `else if`?

see-quick reviewed (commented):
Looks impressive, just a few questions.

scholzj commented on api/src/main/java/io/strimzi/api/kafka/Crds.java:
```diff
@@ -269,6 +269,14 @@ public static CustomResourceDefinition kafka() {
         return client.customResources(CustomResourceDefinitionContext.fromCrd(crd(Kafka.class, Constants.V1ALPHA1)), Kafka.class, KafkaList.class, DoneableKafka.class);
     }
 
+    public static MixedOperation<Kafka, KafkaList, DoneableKafka, Resource<Kafka, DoneableKafka>> kafkaV1Beta1Operation(KubernetesClient client) {
+        return client.customResources(CustomResourceDefinitionContext.fromCrd(crd(Kafka.class, Constants.V1BETA1)), Kafka.class, KafkaList.class, DoneableKafka.class);
+    }
+
+    public static MixedOperation<Kafka, KafkaList, DoneableKafka, Resource<Kafka, DoneableKafka>> kafkaV1Beta2Operation(KubernetesClient client) {
+        return client.customResources(CustomResourceDefinitionContext.fromCrd(crd(Kafka.class, Constants.V1BETA2)), Kafka.class, KafkaList.class, DoneableKafka.class);
+    }
```
What is the difference for these two? Will each of these watch for another version resource? Will we need to do everything twice in the operators?

scholzj reviewed (commented):
I left some comments. But I think I'm less concerned about the code than about the plan. We need to be at CRD v1 before Kube 1.22. So we are less than 6 months away. So can we outline the time plan how to get away with the `listeners` OneOf field issue?

ppatierno commented on api/src/main/java/io/strimzi/api/kafka/model/KafkaClusterSpec.java:
```diff
@@ -146,6 +147,7 @@ public void setLogging(Logging logging) {
         this.logging = logging;
     }
 
+    @PresentInVersions("v1alpha1-v1beta1")
```
What's the syntax for the value of this annotation? It's an interval or a list? or something different?

tombentley replied:
It's described here: https://github.com/strimzi/strimzi-kafka-operator/pull/3653/files#diff-762d04edf28bcda2d851e559e33f5ea4R93. In this case it's an interval with inclusive ends (we don't actually know whether there is a v1alpha2, we just compute inclusion based on the endpoints). It can also be an open-ended interval (e.g. `v1beta2+`), or a single version (`v1beta1`).

ppatierno replied:
Thanks, got it!

ppatierno commented on crd-generator/src/main/java/io/strimzi/crdgenerator/ApiVersion.java:
```diff
@@ -0,0 +1,117 @@
+            }
+            minor = parseShort(matcher.group(4));
+        } else {
+            ab = 2;
```
can we think more about using an enum as Alpha, Beta and Stable instead of short values 0, 1 and 2 (that anyway could be the corresponding underlying values of the enum)

tombentley replied:
There's no benefit to using an enum because these things are entirely internal to `ApiVersion`. The intention is that people use the String form (v1alpha1) rather than using a constructor which takes actual values. Using constants would actually make the `compareTo` code harder to read imho, but I'll do it anyway.

ppatierno replied:
Well, I don't think that the benefit has to be for people using it only but even for developers maintaining it. I got the values now or at least you could document them so no need to update but I think that using enums don't hurt.

tombentley commented on api/pom.xml:
```diff
@@ -164,6 +169,76 @@
+                        <configuration>
+                            <executable>java</executable>
+                            <arguments>
+                                <argument>-classpath</argument>
+                                <argument>${pom.basedir}${file.separator}target${file.separator}classes${path.separator}${pom.basedir}${file.separator}..${file.separator}crd-generator${file.separator}target${file.separator}crd-generator-${project.version}.jar</argument>
+                                <argument>io.strimzi.crdgenerator.CrdGenerator</argument>
+                                <argument>--label</argument><argument>app:strimzi</argument>
+                                <argument>--label</argument><argument>strimzi.io/crd-install:true</argument>
+                                <argument>--target-kube</argument><argument>1.16+</argument>
+                                <argument>--crd-api-version</argument><argument>v1beta1</argument>
+                                <argument>--api-versions</argument><argument>v1beta1,v1beta2</argument>
+                                <argument>--storage-version</argument><argument>v1beta1</argument>
+                                <argument>--yaml</argument>
+                                <argument>io.strimzi.api.kafka.model.Kafka=src/test/resources/io/strimzi/api/kafka/model/040-Crd-kafka-v1beta1-v1beta2-store-v1beta1.yaml</argument>
+                            </arguments>
+                        </configuration>
+                    </execution>
+                    <execution>
+                        <id>generate-crd-evolution-3</id>
+                        <phase>process-classes</phase>
+                        <goals>
+                            <goal>exec</goal>
+                        </goals>
+                        <configuration>
+                            <executable>java</executable>
+                            <arguments>
+                                <argument>-classpath</argument>
+                                <argument>${pom.basedir}${file.separator}target${file.separator}classes${path.separator}${pom.basedir}${file.separator}..${file.separator}crd-generator${file.separator}target${file.separator}crd-generator-${project.version}.jar</argument>
+                                <argument>io.strimzi.crdgenerator.CrdGenerator</argument>
+                                <argument>--label</argument><argument>app:strimzi</argument>
+                                <argument>--label</argument><argument>strimzi.io/crd-install:true</argument>
+                                <argument>--target-kube</argument><argument>1.16+</argument>
+                                <argument>--crd-api-version</argument><argument>v1beta1</argument>
+                                <argument>--api-versions</argument><argument>v1beta1,v1beta2</argument>
+                                <argument>--storage-version</argument><argument>v1beta2</argument>
+                                <argument>--yaml</argument>
+                                <argument>io.strimzi.api.kafka.model.Kafka=src/test/resources/io/strimzi/api/kafka/model/040-Crd-kafka-v1beta1-v1beta2-store-v1beta2.yaml</argument>
+                            </arguments>
+                        </configuration>
+                    </execution>
```
I'll add another yaml with just Kafka `v1beta2` and using CRD `v1`.

tombentley replied:
Thanks, this was a really helpful comment. There were some gotchas with v1 which I hadn't noticed and were caught by actually, _y'know_, testing v1.

tombentley commented:
I think I've addressed most of the comments from last time. I've also extended the `ApiEvolutionCrdIT` to cover upgrading all the way to `v1` of the CRD API. This requires bumping the version of fabic8 to pick up one which supported that. My understanding from @stanlyDoge is that that version has some other problems which are being fixed in upstream fabric8. So that means this PR is currently not mergeable. However it proves how we can migrate to a structural schema by introducing `v1beta2` apis and removing `v1alpha1` and `v1beta1`. It would be necessary to do this before Kubernetes 1.21 was released. Probably makes sense to discuss this in detail in the next community meeting on Thursday.

stanlyDoge commented:
Yeah, we need to wait for a stable fabric8 version. For now, it seems like 4.11.2 might be the one. Follow https://github.com/strimzi/strimzi-kafka-operator/pull/3553 for more details.

tombentley commented:
@scholzj:

1. Yes, the test part of this does depend on Fabric8 4.11, since that's the first version which supports both api versions of CRD. In theory we could split it up and merge the test later, but let's see how the fabric8 folks answer [my question](https://github.com/tombentley) about when a fixed version, required by #3553, might be released. Once that exists we'd have a conflict between this PR and #3553. I guess we would either merge that first or make this pass the CI.

2. I think it _would_ supersede both those other PRs.

tombentley commented:
Fabric8 release might be as soon as today. @stanlyDoge if it _is_ today, could to take a look at #3553 next week?

scholzj commented:
@tombentley I merged the Fabric8 PR ... so I guess this will now need some rebase or something. ALso it looks like the DCO failed for some reason. Next week should be fine of course.

tombentley commented:
/azp run system-tests

scholzj commented:
/azp run acceptance

scholzj commented:
@samuel-hawker @ppatierno @see-quick Could you please do another pass through this? It would be good to merge this to be able to move forward towards 0.20.0.

scholzj commented:
/azp run regression

tombentley commented:
Just to be clear: this doesn't _have_ to be included in 0.20.0, but it could be. If it wasn't merged for 0.20, it would make sense to merge it soon afterwards.

samuel-hawker reviewed (approved):
LGTM
